doc/cluster_design.md

```diff
@@ -0,0 +1,137 @@
+- `model.n_clusters` is the number of patterns after clustering.
+- `my_cluster_model` is the trained cluster model.
+- `run_pretrain`  is used to determine if autoencoder pretrain needs to be run, default true.
+- `model.existed_pretrain_model` is used to specify an existing pretrain_model
+- `output_table` is the cluster result for input_table, which is adding the `group_id` column predicted by the cluster model to the input_table.
+
+## clusterModel Details
+<img src="figures/cluster_model_train_overview.png">
+
+The below figure demonstrates overall workflow for clusterModel train. This figure includes two parts, the pretrian autoencode model and the cluster model are included.
+1. First, the former is used to train a pretrain model. The `model.encode_units` describes the layer structure of the encoder of the autoencoder network. We only use the output of the trained encode layer (10000*7) as the input to the clustering model. 
+2. Then, the clustering model starts training, randomly initializes weights and multiple iterations, generates clustering models.
+3. Finally, the overall train process ultimately outputs an unsupervised clustering model.
+
+
+## Implement Details
+- sqlflow_models/clusterModel.py
+
+```python
+class  clusterModel(tf.keras.Model):
+
+	def pre_train(dataset):
+			...
+			self.autoencoder.fit(dataset)
+		pretrainmodel.save(‘/tmp/pretrain.h5’）
+
+	def target_distribution():
+		...
+
+	def  cluster_train_loop():
+		for ite in range(int(maxiter)):
+			if ite % update_interval == 0:
+				q = model.predict(x, verbose=0)
+				p = target_distribution(q)  # update the auxiliary target distribution p
+				y_pred = q.argmax(1)
+			idx = index_array[index * batch_size: min((index+1) * batch_size, x.shape[0])]
+				loss = model.train_on_batch(x=x[idx], y=p[idx])
+			 index = index + 1 if (index + 1) * batch_size <= x.shape[0] else 0
+```
+
+- template_tf.go
+```python
+if 'pre_train' is in classifier:
+	classifier.pre_train(InputDataSet)
+if 'cluster_train_loop' is in classifier:
+	classifier.cluster_train_loop(InputDataSet)
+
+```
+
+## Note
+
+- The user can choose whether to run pre_train before the cluster model, ie run_pretrain=true. The user can also choose to load the already trained model by loading the existed_pretrain_model.
+Therefore, there are four cases in total:
+1.  run_pretrain = true & Using model.existed_pretrain_model = None：
+Autoencoder Pretrain + Random initialization weights for cluster. (Note that model.encode_units `is work` at this time.)
+2.  run_pretrain = true & Using model.existed_pretrain_model = existed_pretrain_model：
+existed_pretrain_model Pretrain+ Random initialization weights for cluster. (Note that model.encode_units `is not work` at this time.)
+3.  run_pretrain = false & Using model.existed_pretrain_model = None: 
+Random initialization weights for cluster. (Note that model.encode_units `is not work` at this time.)
```

Comment:
You mean “does not work”? Same in other places. Also did you mean to use double quotes instead?

Reply:
The `USING` statement have a higher precedence than  pre_train=True in WITH statement. Since model.encode_units sets the pre_train part of the autoencoder network, the pre_train statement does not work when the using statement exists, so encode_units does not work.

Reply:
I understand that. I was just asking you to revisit this part of the grammar. It should be “does” instead of “is”. And the. Switch to use double quotes.

Reply:
OK~~thx！